Add api trade simulation scripts

ghubstan opened the pull request:
Two regtest trading simulation scripts are contained in this change:

- `trade-simulation.sh` goes through the steps of creating country based F2F payment accounts for Bob & Alice, plus each step of a trade from `createoffer` to completion.
  
- `limit-order-simulation.sh` shows a way to trigger creation of an offer when a limit price is reached.
  
Each script prints CLI commands and output as they are run, providing a basic tutorial for some of the api's features.

Examples:

- `trade-simulation.sh`          

  Simulate the entire trade protocol between Bob (taker) & Alice (maker),
  where Alice buys 0.1 BTC from Bob, paying in Renminbi (CYN).

  Note the script takes a country code (CN) not a currency code, so the
  script can create the appropriate country based face to face payment accounts.

  `$ apitest/scripts/trade-simulation.sh -d buy -c cn -m 0.0 -a 0.1`

- `limit-order.sh`          

  Create a CAD/BUY 0.1 BTC order at mkt price margin of 0.0% if price falls to
  or below 47900 CAD.

  Note the script takes a country code (CA) not a currency code, so the script
 can create the appropriate country based face to face  payment accounts.

  `$ apitest/scripts/limit-order-simulation.sh -l 47900 -d buy -c CA -m 0.0 -a 0.1`

  Create a USD/SELL 0.1 BTC order at mkt price margin of 0.0% if price rises to
  or above 37200 USD.

  `$ apitest/scripts/limit-order-simulation.sh -l 37200 -d sell -c US -m 0.0 -a 0.1
`

Both scripts depend on functions contained in supporting `bash` and `python3` scripts.

https://github.com/bisq-network/bisq/pull/5084 should be merged before this one.

ghubstan committed:
Add line break

ghubstan committed:
Make opt description generic

ghubstan committed:
Add method help docs

ghubstan committed:
Add method help docs

ghubstan committed:
Fix help text typos

ghubstan committed:
Add --memo opt to help text

ghubstan committed:
Add method help docs

ghubstan committed:
Fix typo

ghubstan committed:
Add method help doc

ghubstan committed:
Explain example

ghubstan committed:
Add method help doc

ghubstan committed:
Add method help docs

ghubstan committed:
Explain example

ghubstan committed:
Keep CLI method enum members in alphabetical order

ghubstan committed:
Merge branch 'master' into 01-add-method-help-docs

ghubstan committed:
Add api trade simulation scripts

Two regtest trading simulation scripts are contained in this change:

- trade-simulation.sh goes through the steps of creating F2F payment
  accounts for Bob & Alice, and each step of a trade from createoffer to
  completion.

- limit-order-simulation.sh shows one way to trigger creation of an offer
  when a limit price is reached.

Each script prints CLI commands just before they are run.

Both scripts depend on functions contained in supporting bash and python3
scripts.

Examples:

trade-simulation.sh

  Simulate the entire trade protocol between Bob (taker) & Alice (maker),
  where Alice buys 0.1 BTC from Bob, paying in Renminbi (CYN).

  Note the script takes a country code (CN) not a currency code, so the
  script can create the appropriate country based face to face payment accounts.

  $ apitest/scripts/trade-simulation.sh -d buy -c cn -m 0.0 -a 0.1

limit-order.sh

  Create a CAD/BUY 0.1 BTC order at mkt price margin of 0.0% if price falls to
  or below 47900 CAD.

  Note the script takes a country code (CA) not a currency code, so the script
  can create the appropriate country based face to face  payment accounts.

  $ apitest/scripts/limit-order-simulation.sh -l 47900 -d buy -c CA -m 0.0 -a 0.1

  Create a USD/SELL 0.1 BTC order at mkt price margin of 0.0% if price rises to
  or above 37200 USD.

  $ apitest/scripts/limit-order-simulation.sh -l 37200 -d sell -c US -m 0.0 -a 0.1

ghubstan committed:
Fix bash bugs & style problems for codacy

ghubstan committed:
Add env requirement and usage comments

ghubstan committed:
Make price request frequency configurable

Added a -w option to  allow user to control the frequency of price requests.
The minimum value = 20s, default value = 120s.

sqrrm commented on apitest/scripts/trade-simulation.sh:
```diff
@@ -0,0 +1,228 @@
+# Complete the trade on the seller side.
+if [ "${DIRECTION}" = "BUY" ]
+then
+	printdate "BOB ${BOB_ROLE}:  Closing trade by keeping funds in Bisq wallet."
```
Indent

ghubstan replied:
Fixed in commit 4eabb9804eb979db1c64ee336e74c2ec97b0f70d.

sqrrm commented on apitest/scripts/trade-simulation.sh:
```diff
@@ -0,0 +1,228 @@
+    CMD="$CLI_BASE --port=${BOB_PORT} keepfunds --trade-id=${OFFER_ID}"
+    printdate "BOB CLI: ${CMD}"
+else
+	printdate "ALICE (taker):  Closing trade by keeping funds in Bisq wallet."
```
Indent

ghubstan replied:
Fixed in commit 4eabb9804eb979db1c64ee336e74c2ec97b0f70d.

sqrrm commented on core/src/main/resources/help/canceloffer-help.txt:
```diff
@@ -0,0 +1,25 @@
+
+DESCRIPTION
+-----------
+Cancel an existing offer.  The offer will be removed from other Bisq users' offer views.
```
Might be nice to add that fees paid will be forfeited.

ghubstan replied:
Fixed in commit 71ffa0f9979cd22e91c8c6a5e3cca21fa677505b.

sqrrm commented on core/src/main/resources/help/getmyoffer-help.txt:
```diff
@@ -0,0 +1,25 @@
+
+DESCRIPTION
+-----------
+List one of your existing offer's details.
```
```suggestion
List one of your existing offers' details.
```

ghubstan replied:
Nice catch!  Fixed in commit 86758414b.

sqrrm commented on core/src/main/resources/help/getoffer-help.txt:
```diff
@@ -0,0 +1,26 @@
+DESCRIPTION
+-----------
+List an existing offer's details.  The offer cannot be one of your own, and be
+available for the taking, or it will not be viewable.
```
This wording is a bit confusing. Better separate to something like "Must not be your own. Must be available for taking."

ghubstan replied:
Fixed in commit fce3aab40.

sqrrm commented on core/src/main/resources/help/keepfunds-help.txt:
```diff
@@ -0,0 +1,31 @@
+Bisq wallet.  This step may not seem necessary from the buyer's perspective, but it is
+necessary for correct transition of a trade's state to CLOSED, within the Bisq server.
+
+The alternative way to close out the trade is to sen the received BTC to an external
```
```suggestion
The alternative way to close out the trade is to send the received BTC to an external
```

ghubstan replied:
Fixed in commit 031b18edf.

sqrrm commented on core/src/main/resources/help/sendbsq-help.txt:
```diff
@@ -0,0 +1,38 @@
+--tx-fee-rate
+        An optional transaction fee rate (sats/byte) for the transaction.  The user is
+        responsible for choosing a fee rate that will be accepted by the network in a
+        reasonable amount of time, and the fee rate must be greater than 0.
```
```suggestion
        reasonable amount of time, and the fee rate must be greater than 1.
```

ghubstan replied:
Fixed in commit 67eed01d7.

sqrrm commented on core/src/main/resources/help/sendbtc-help.txt:
```diff
@@ -0,0 +1,51 @@
+--tx-fee-rate
+        An optional transaction fee rate (sats/byte) for the transaction.  The user is
+        responsible for choosing a fee rate that will be accepted by the network in a
+        reasonable amount of time, and the fee rate must be greater than 0.
```
```suggestion
        reasonable amount of time, and the fee rate must be greater than 1.
```

ghubstan replied:
Fixed in commits 11096d5c8 and 72731e5c0.

sqrrm reviewed (changes requested):
Some minor changes.

I ran the scripts and I find the resulting orders strange, the price doesn't seem to correlate to the price feed, would like to understand what's happening there.

dmos62 commented on apitest/scripts/limit-order-simulation.sh:
```diff
@@ -0,0 +1,149 @@
+#
+#     `$ apitest/scripts/limit-order-simulation.sh -l 40000 -d buy -c fr -m 3.00 -a 0.125`
+#
+#  Script options:  -l <limit-price> -d <direction> -c <country-code> -m <mkt-price-margin(%)> - f <fixed-price> -a <amount(btc)> [-w <price-poll-interval(s)>]
```
`- f <fixed-price>` -> `-f <fixed-price>`

ghubstan replied:
Fixed in commit 73c6b3d96.

dmos62 commented on apitest/scripts/trade-simulation-env.sh:
```diff
@@ -0,0 +1,219 @@
+export ARBITRATOR_PORT=9997
+export ALICE_PORT=9998
+export BOB_PORT=9999
+export F2F_ACCT_FORM="f2f-acct.json"
```
`export` is for putting these variables in child processes' environments, but I don't think these variables are used by child processes in this case. If these variables are only used by other scripts by having this one `source`d, then the `export` prefixes aren't necessary.

ghubstan replied:
Fixed in commit 8b4938a9e.

ghubstan replied:
Turns out codacy requires these vars be exported, or it fails the codacy analysis, and cannot be merged.  It may be fixed by tweaking the codacy config, but that won't happen immediately.

dmos62 commented on apitest/scripts/trade-simulation-utils.sh:
```diff
@@ -0,0 +1,174 @@
+}
+
+genbtcblocks() {
+	NUM_BLOCKS=$1
```
It's safer to double quote these variable references (`NUM_BLOCKS="$1"`). See https://tldp.org/LDP/abs/html/quotingvar.html It's a quirk of Bash. Applies to other variable references in the PR as well.

ghubstan replied:
Fixed in commit 6c322d4aae1b273ce81242816d55efc1a1f47ff8.

dmos62 commented on apitest/scripts/trade-simulation-utils.sh:
```diff
@@ -0,0 +1,174 @@
+
+getbtcoreaddress() {
+    CMD="bitcoin-cli -regtest  -rpcport=19443 -rpcuser=apitest -rpcpassword=apitest getnewaddress"
+    NEW_ADDRESS=$(${CMD})
```
`${` is inconsistent with some other variable references. It's usually used when you have something like `${var}postifx` (you couldn't do `$varpostfix`). `$($CMD)` should be enough here. Though either is fine, whatever you think is more readable, it's just that it would be nice if the choice were consistent.

ghubstan replied:
Fixed in commit cb96f27f1.

dmos62 commented on apitest/scripts/trade-simulation.sh:
```diff
@@ -0,0 +1,228 @@
+printdate "Bob and Alice create their face to face ${COUNTRY_CODE} payment accounts."
+CMD="${CLI_BASE} --port=${BOB_PORT} createpaymentacct --payment-account-form=${APITEST_SCRIPTS_HOME}/${F2F_ACCT_FORM}"
+printdate "BOB CLI: ${CMD}"
+CMD_OUTPUT=$(createpaymentacct "${CMD}")
```
These references are a bit more complicated to quote because of nested variable references. Escaping the inner quotes should do the trick: `CMD_OUTPUT="$(createpaymentacct \"${CMD}\")"`. Readability isn't great. Alternatively, at the risk of too much vertical verbosity, you could split the inner part into a new variable: `tmp="createpaymentacct $CMD"`.

ghubstan replied:
Will you tell me why I need to make this change?  It is working.  
What is wrong with  `CMD_OUTPUT=$(createpaymentacct "${CMD}")`?

dmos62 replied:
It's the same variable referece quoting issue I refer to in other comments. It prevents reinterpretation of some special characters, prevents word splitting. See examples here: https://tldp.org/LDP/abs/html/quotingvar.html

dmos62 commented on apitest/scripts/trade-simulation.sh:
```diff
@@ -0,0 +1,228 @@
+
+# Source the env and some helper functions.
+. "${APITEST_SCRIPTS_HOME}/trade-simulation-env.sh"
+. "${APITEST_SCRIPTS_HOME}/trade-simulation-utils.sh"
```
The longform `source` might be preferable to the shortform `.` due to readability.

ghubstan replied:
Fixed in commit 0c06abdc5.

ghubstan committed:
Fix indentation

For suggested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560898873

ghubstan committed:
Explain that paid tx fees are forfeited

For suggested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560901896

ghubstan committed:
Fix punctuation

For suggested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560902858

ghubstan committed:
Improve description

For suggested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560905251

ghubstan committed:
Fix typo

For requested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560906396

ghubstan committed:
Fix tx-fee-rate opt description

For requested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560907371

ghubstan committed:
Add parens around (sats/byte)

ghubstan committed:
Fix tx-fee-rate opt description

For requested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560906913

ghubstan committed:
Fix typo, group mutually exclusive opts around '||'

ghubstan committed:
Put $N variable references in double quotes

For requested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560938698

ghubstan committed:
Replace shortform '. with longform 'source'

For requested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560958418

ghubstan committed:
Do not export vars from lowest level child script

For requested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560932201

ghubstan committed:
Remove uneeded ${curlies}

For requested change
https://github.com/bisq-network/bisq/pull/5093#discussion_r560941709

ghubstan committed:
Remove api dev/test log statement

This server log output was intended as an aid to api devs, but
is no longer needed after the change to posix-sytle method opts
with self explanatory labels (replacing the ambiguous positional
CLI method opts).

dmos62 commented on apitest/scripts/trade-simulation-utils.sh:
```diff
@@ -137,14 +137,14 @@ getnewpaymentacctcurrency() {
 createpaymentacct() {
-    CMD=$1
+    CMD="$1"
     CMD_OUTPUT=$(${CMD})
```
You missed some variable refefences: `"$(${CMD})"`

ghubstan committed:
Print current market price just before offer is created

This may help confirm offer.price correctness in periods of BTC
price stability.

ghubstan committed:
Add new api CLI method 'getbtcprice'

The server impl was there, but it is now needed by the trading
sim scripts (CLI) to get the price from the Bisq server instead
of the feed.  (The server does not request prices more than
once a minute.)

ghubstan committed:
Get current mkt price from server, not directly from feed

ghubstan committed:
Remove deprecated python script

sqrrm reviewed (dismissed):
utACK

ghubstan committed:
Re-add exports to make codacy happy

ghubstan committed:
Enclose $vars in double quotes

sqrrm reviewed (dismissed):
utACK

ghubstan commented:
Force travis build (jar download problem).

ghubstan committed:
Run an async price feed request when CLI needs market price

The CLI was receiving stale, cached market prices from the feed service.

ghubstan committed:
Log price feed request warnings, do not throw to CLI

The price feed service throws PriceRequestExceptions when switching
currencies, log those exceptions as warnings in the server and don't
pass them up to the CLI.

ghubstan commented:
Note on commit b2d8faf:  This fixes a bug in the core api that was throwing the `PriceRequestException` exception below to the CLI when switching currencies, despite a price being available.
```
PriceFeedService: Exception at applyPriceToConsumer for currency CNY. priceProvider=https://price.bisq.wiz.biz/. Exception=java.lang.IllegalStateException: Stream is already completed, no further calls are allowed 
Jan-20 20:27:01.254 [BisqDaemonMain] WARN  b.c.a.CorePriceService: Exception at applyPriceToConsumer for currency CNY. priceProvider=https://price.bisq.wiz.biz/. Exception=java.lang.IllegalStateException: Stream is already completed, no further calls are allowed bisq.core.provider.price.PriceRequestException: Exception at applyPriceToConsumer for currency CNY. priceProvider=https://price.bisq.wiz.biz/. Exception=java.lang.IllegalStateException: Stream is already completed, no further calls are allowed
	at bisq.core.provider.price.PriceFeedService.applyPriceToConsumer(PriceFeedService.java:387)
	at bisq.core.provider.price.PriceFeedService.lambda$request$0(PriceFeedService.java:174)
	at bisq.core.provider.price.PriceFeedService$1.lambda$onSuccess$0(PriceFeedService.java:417)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)

```

sqrrm reviewed (approved):
utACK
